### PR TITLE
Conditional option create

### PR DESCRIPTION
**Pull request by v1ctor**
Hi there. 
I propose this enhancement because sometimes it's really convenient.
Hope that it can be useful for another people.

### CODE REVIEW

**Comment by danieldietrich** [comment]
Hi @v1ctor, thank you for your proposal!
Yes, this might come handy. I will comment the code in a minute...

File: src/main/java/javaslang/control/Option.java [hunk 1]
```diff
@@ -53,6 +53,19 @@
     }
 
     /**
+     *
+     * Creates {@code Some} of suppliers value if condition is true, or {@code None} in other case
+     *
+     * @return return {@code Some} of supplier's value if condition is true, or {@code None} in other case
+     */
+    static <T> Option<T> when(boolean condition, Supplier<T> supplier) {
```
Comment (danieldietrich) [inline src/main/java/javaslang/control/Option.java]: Let's use `Supplier<? extends T>`. There are situations where it is necessary.

Also we do null checks in order to provide user-friendly messages, e.g. `Objects.requireNonNull(supplier, "supplier is null");`

File: src/main/java/javaslang/control/Option.java [hunk 2]
```diff
@@ -53,6 +53,19 @@
...
+        if (!condition) {
+            return none();
+        }
+        return of(supplier.get());
```
Comment (danieldietrich) [inline src/main/java/javaslang/control/Option.java]: I would to a `condition ? of(supplier.get()) : none();` here but that is a matter of taste. Using your guard method is ok!

**Comment by v1ctor** [comment]
Thank you. I have fixed it in accordance to your comments.

**Comment by danieldietrich** [comment]
Very nice, many thanks!

**Pull request merged**